### PR TITLE
fix #17764, type problem due to object_id collision after `workspace()`

### LINKED ISSUE #17764: workspace() and LastMain regression
@maleadt and I uncovered a regression in the `workspace()` behavior. Compare the behavior of calling the constructors of `Foo` and `Bar` after `workspace()`:

```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-rc0+110 (2016-08-01 16:59 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit e2d2925* (1 day old master)
|__/                   |  x86_64-apple-darwin15.5.0

julia> type Foo
          foo::Vector{Float32}
       end

julia> Foo([1.])
Foo(Float32[1.0])

julia> type Bar
          foo::String
       end

julia> Bar("bar")
Bar("bar")

julia> workspace()

julia> type Foo
          foo::Vector{Float32}
       end
WARNING: Method definition (::Type{Main.Foo})(Array{Float32, 1}) in module Main at REPL[1]:2 overwritten in module Main at REPL[7]:2.

julia> Foo([1.])
LastMain.Foo(Float32[1.0])

julia> type Bar
          foo::String
       end
WARNING: Method definition (::Type{Main.Bar})(String) in module Main at REPL[3]:2 overwritten in module Main at REPL[10]:2.

julia> Bar("bar")
Bar("bar")
```

However, on 0.4.6:

```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.6 (2016-06-19 17:16 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-apple-darwin13.4.0

julia> type Foo
          foo::Vector{Float32}
       end

julia> Foo([1.])
Foo(Float32[1.0f0])

julia> type Bar
          foo::String
       end
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0

julia> Bar("bar")
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0
Bar("bar")

julia> workspace()

julia> type Foo
          foo::Vector{Float32}
       end

julia> Foo([1.])
Foo(Float32[1.0f0])

julia> type Bar
          foo::String
       end
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0

julia> Bar("bar")
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0
WARNING: Base.String is deprecated, use AbstractString instead.
  likely near no file:0
Bar("bar")
```

Also, on 0.5,
```
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-rc0+110 (2016-08-01 16:59 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit e2d2925* (1 day old master)
|__/                   |  x86_64-apple-darwin15.5.0

julia> type Foo end

julia> workspace()

julia> type Foo end
WARNING: Method definition (::Type{Main.Foo})() in module Main at REPL[1]:1 overwritten in module Main at REPL[3]:1.

julia> @which Foo()
LastMain.Foo() at REPL[3]:1

julia> typeof(Foo()) == Main.Foo
true

julia> typeof(Foo()) == LastMain.Foo
false
```

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

File: base/docs/helpdb/Base.jl [hunk 1]
```diff
@@ -6416,7 +6416,7 @@ ctranspose!
 """
     object_id(x)
 
-Get a unique integer id for `x`. `object_id(x)==object_id(y)` if and only if `is(x,y)`.
+Get a hash value for `x` based on object identity. `object_id(x)==object_id(y)` if `x === y`.
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: run `make docs` and commit the rst, or it'll confuse someone else as an unrelated change to whatever they're working on

**Comment by tkelman** [comment]
This broke SHA.jl. Not good.

**Comment by vtjnash** [comment]
build link?

**Comment by tkelman** [comment]
https://travis-ci.org/staticfloat/SHA.jl/jobs/149759702

```
INFO: Testing SHA
ERROR: LoadError: MethodError: no method matching digestlen(::Type{SHA.SHA3_384_CTX})
Closest candidates are:
  digestlen(::Type{SHA.SHA1_CTX}) at /home/travis/.julia/v0.6/SHA/src/types.jl:74
  digestlen(::Type{SHA.SHA2_224_CTX}) at /home/travis/.julia/v0.6/SHA/src/types.jl:75
  digestlen(::Type{SHA.SHA3_224_CTX}) at /home/travis/.julia/v0.6/SHA/src/types.jl:76
  ...
 in describe_hash(::Type{SHA.SHA3_384_CTX}) at /home/travis/.julia/v0.6/SHA/test/runtests.jl:97
 in (::##1#2)(::Pair{Any,DataType}) at ./<missing>:0
 in next(::Base.Generator{Dict{Any,DataType},##1#2}, ::Int64) at ./generator.jl:26
 in collect_to!(::Array{String,1}, ::Base.Generator{Dict{Any,DataType},##1#2}, ::Int64, ::Int64) at ./array.jl:307
 in collect_to_with_first!(::Array{String,1}, ::String, ::Base.Generator{Dict{Any,DataType},##1#2}, ::Int64) at ./array.jl:294
 in collect(::Base.Generator{Dict{Any,DataType},##1#2}) at ./array.jl:275
 in include_from_node1(::String) at ./loading.jl:426
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
while loading /home/travis/.julia/v0.6/SHA/test/runtests.jl, in expression starting on line 100
```

**Comment by vtjnash** [comment]
possibly the same issue as #17809, which was pre-existing

**Comment by tkelman** [comment]
Is reverting this PR on release-0.5 a valid short-term fix? How did our tests not identify the problem here?

**Comment by StefanKarpinski** [comment]
What's the status of this?

**Comment by tkelman** [comment]
fixed by https://github.com/JuliaLang/julia/pull/17820 which was backported to release-0.5, and binaries of rc1+1 were posted on the mailing list (and replaced 0.5-latest with it for travis and appveyor)

**Pull request merged**